oak_crypto/src/aead.rs

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{schema::AeadEncryptedMessage, MessageWithAssociatedData};
+use aes_gcm::{aead::AeadInPlace, Aes256Gcm, Key, KeyInit, Nonce};
+use anyhow::anyhow;
+
+/// AES-256-GCM ID.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-authenticated-encryption-wi>
+pub(crate) const AEAD_ID: u16 = 0x0002;
+/// Represents `N_k` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const AEAD_ALGORITHM_KEY_SIZE_BYTES: usize = 32;
+/// Represents `N_n` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const AEAD_NONCE_SIZE_BYTES: usize = 12;
+/// Convenience type for representing an AEAD key.
+pub(crate) type AeadKey = [u8; AEAD_ALGORITHM_KEY_SIZE_BYTES];
+/// Convenience type for representing an AEAD nonce.
+pub(crate) type AeadNonce = [u8; AEAD_NONCE_SIZE_BYTES];
+
+/// Implementation of Authenticated Encryption with Associated Data (AEAD).
+/// <https://datatracker.ietf.org/doc/html/rfc5116>
+pub(crate) struct Aead {}
+
+impl Aead {
+    /// Encrypts message with associated data using AES-GCM encryption scheme.
+    /// Returns a serialized [`AeadEncryptedMessage`] proto.
+    ///
+    /// Note: the corresponding associated data is NOT encrypted.
+    pub(crate) fn encrypt(
+        secret_key: &AeadKey,
+        nonce: &AeadNonce,
+        plaintext: &MessageWithAssociatedData,
+    ) -> anyhow::Result<AeadEncryptedMessage> {
+        let cipher = Aes256Gcm::new(Key::<Aes256Gcm>::from_slice(secret_key));
+
+        // Encrypt message.
+        let mut encrypted_message = plaintext.message.to_vec();
+        cipher
+            .encrypt_in_place(
+                Nonce::from_slice(nonce),
+                &plaintext.associated_data,
+                &mut encrypted_message,
+            )
+            .map_err(|error| anyhow!("couldn't encrypt data: {}", error))?;
+
+        let encrypted_message = AeadEncryptedMessage {
+            encrypted_message,
+            associated_data: plaintext.associated_data.to_vec(),
+        };
+        Ok(encrypted_message)
+    }
+
+    /// Decrypts and authenticates `serialized_encrypted_message` using AES-GCM encryption scheme.
```

Comment:
`serialized_encrypted_message` does not appear in the list of param

Reply:
Done